applications/Fennel_Protocol.md

```diff
@@ -0,0 +1,274 @@
+
+### Milestone 2 — Additional features
+
+* **Estimated duration:** 1 months
+* **FTE:**  3
+* **Costs:** 15,000 USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | Unlicense |
+| 0b. | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can (for example) spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
+| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| 0d. | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 0e. | Article | We will publish an **article** that describes key generation and management with Fennel Protocol. |
+| 1. | Keypair Creation | We will integrate system-level key management libraries to provide a collection of non-blockchain key sources. We'll include AES 256 and RSA 4096 at this point. |
+| 2. | Public Key Transmission | We will develop a Substrate module that contains public key management extrinsics for broadcasting public keys' existence and locations as a signal. |
+| 3. | Public Key Retrieval | We will add an extrinsics set to the Key Management module allowing for simple retrieval of public keys given their location.<br />This will necessarily include a mechanism for verifying that the key retrieved is the key requested. |
+| 4. | Encrypted Communciation Channel Negotiation | We will create a Substrate module with extrinsics used to negotiate a three-way key handshake between two parties through cooperation of on- and off-chain compute. |
+
+### Milestone 3 — Additional features
+
+* **Estimated duration:** 1 months
+* **FTE:**  3
+* **Costs:** 20,000 USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | Unlicense |
+| 0b. | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can (for example) spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
+| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| 0d. | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 0e. | Article | We will publish an **article** overviewing using Fennel Protocol in applications for secure off-chain data-centric messaging and secure on-chain signal mechanisms. |
+| 1. | Peer Retrieval | We will construct a Substrate-resident method for applications exposing the same services to identify and locate one another through on-chain signals. |
```

Comment:
Can you explain this a bit more? It sounds like this is for different Whiteflag instances to find one another? What would be the purpose of that?

Reply:
This is for applications _using_ the Whiteflag protocol to find one another. For example, if you've got two different platforms using Fennel's Whiteflag mechanism, the peer retrieval functionality lets them identify each other and create private, off-chain channels between themselves using only publicly-available information.

Reply:
Can you explain how this works? What are these on-chain signals, and how does one find them?

Reply:
Signals on Fennel Protocol are an abstraction for calls to a set of extrinsics that interact with on-chain storage and events to broadcast and query finite pieces of information. In this case, an application that wants to use them will have already set up an identity on the Fennel Protocol network, then will use those extrinsics to store a public announcement (the actual signal) that it's the identity for a particular application that implements a service or protocol. From there, given that identity and the chain-parallel mechanisms I mention in the other section you commented on, another application or service can query storage (through another extrinsic included in this deliverable) and identify others implementing the protocol they need to communicate with and negotiate a direct connection without exposing their IP address to the entire network.

Reply:
I should also comment that when we say "signal", there are Fennel-specific signals and Whiteflag signals. I've described Fennel-specific in the last comment. The Whiteflag signals are implemented identically, but are intended for more human-level communication (reporting natural disasters,sending information to other groups) as opposed to the service-level communication we see in Fennel signals.